R/experiment.R

```diff
@@ -555,215 +701,538 @@ Experiment <- R6::R6Class(
+                }, simplify=FALSE)
+                dplyr::bind_rows(reps, .id = "rep")
+              },
+              dgp_mapply_args, method_mapply_args,
+              future.seed = future.seed, SIMPLIFY = FALSE,
+              future.globals = future.globals,
+              future.packages = future.packages
+            )
+            data.table::rbindlist(results, fill = TRUE) %>%
+              tibble::as_tibble()
+          },
+          "reps+dgps+methods" = {
+            mapply_args <- purrr::cross3(
+              1:n_reps, dgp_params_list, method_params_list
+            )
+            # shuffle the inputs to avoid bad load balancing
+            mapply_args <- mapply_args[sample(1:length(mapply_args))]
+            reps <- sapply(mapply_args, `[[`, 1)
+            dgp_mapply_args <- lapply(mapply_args, `[[`, 2)
+            method_mapply_args <- lapply(mapply_args, `[[`, 3)
+            results <- future.apply::future_mapply(
+              function(dgp_params, method_params) {
+                dgp_name <- dgp_params$dgp_name
+                dgp_params$dgp_name <- NULL
+                data_list <- do.call(dgp_list[[dgp_name]]$generate, 
+                                     dgp_params)
                 method_name <- method_params$method_name
                 param_df <- simplify_tibble(list_to_tibble_row(
                   c(dgp_name = dgp_name, dgp_params, method_params)
                 ))
                 method_params$method_name <- NULL
                 method_params$data_list <- data_list
-                result <- do.call(method_list[[method_name]]$fit, method_params)
+                result <- do.call(method_list[[method_name]]$fit,
+                                  method_params)
                 return(result %>% tibble::add_column(param_df, .before=1))
-              })
-            },
-            future.globals = future.globals,
-            future.packages = future.packages,
-            future.seed = future.seed
-          )
-          results <- dplyr::bind_rows(results)
-          results$rep <- rep(1:n_reps, each = n_dgps*length(method_params_list))
-          results
-        },
-        "reps+methods" = {
-          n_methods <- length(method_params_list)
-          method_params_list <- rep(method_params_list, times = n_reps)
-          results <- future.apply::future_lapply(
-            method_params_list,
-            function(method_params) {
-              purrr::map_dfr(dgp_params_list, function(dgp_params) {
+              },
+              dgp_mapply_args, method_mapply_args,
+              future.seed = future.seed, SIMPLIFY = FALSE,
+              future.globals = future.globals,
+              future.packages = future.packages
+            ) %>%
+              data.table::rbindlist(fill = TRUE) %>%
+              tibble::as_tibble()
+            results$rep <- reps
+            results
+          }
+        )
+      } else {
```

Comment:
Do you think it's possible to get rid of this `else`? I'd ideally like to have just one of these gigantic switches over `parallel_strategy`, because it might be a bit annoying / error-prone to make changes in two places in the future. For example, in the case of "reps" `new_fit_params` could be `NULL` in which case we check for that and avoid the call to `.get_new_method_params()`. 

Also, in that same case we end up calling `.get_new_method_params()` `n_reps` times, but we should only need to call it `length(dgp_params_list)` times I think. If so, it may be better to get our `dgp_params_list` and `method_params_list` sorted before we do any of the parallelization, and then we also shouldn't need the `else`.

Reply:
Good call and easy fix. I'll move the `if` statement to inside the `switch` components.